Adding a writer for Markua 0.10

binaarinen opened the pull request:
This change will add a simple writer to support outputs in Markua 0.10. The 
writer setup is similarly to the Markdown writer to conform to the pandoc 
structure. It is influenced by the Markdown, XWiki and TEI writers and can be 
easily extended to include more functionality in the future.

## Project
The project was done as part of an assignment under Prof. Dr. Farhad D.Mehta at
the Ostschweizer Fachhochschule - OST (university of applied science) by:
    - Tim Wisotzki (tim.wisotzki@ost.ch)
    - Samuel Lemmenmeier (samuel.lemmenmeier@ost.ch)
If possible we would be happy to get some feedback soon, since the project will
conclude with that.

## Functionality
 - convert the Pandoc AST to Markua syntax (as defined in the 
     [Markua Manual](https://leanpub.com/markua/read)) wherever possible 
     (some features are not supported by Markua) to the most general Markua 
     style.
 - common options for Markua (citeproc, bibliography, columns)
 - when the Markua Manual does not include a unique specification, the not
     final [Manual Spec](https://markua.com) was used as a guideline.

## Special considerations
the following assumtions and/or defaults were used for the conversion form AST 
to Markua:
 - attributes are, when available, always added as an attribute-list above the 
    block or around an inline span
 - if the data for footnotes and bibliography are available they are 
    automatically added. 
 - images are inserted as figues and if they have a caption, alt or title, these
    are only inserted in the attribute list.
 - links are always in span
 - divisions will be returned without <div> tags and attribute
 - since Markua doesn't support raw blocks, but they can be used for filters, 
    the are returned in original form, without attributes
 - math will always have the format latex

tarleb commented on src/Text/Pandoc/Writers/Markua.hs:
```diff
@@ -0,0 +1,501 @@
+                                                   thead 
+                                                   tbody 
+                                                   tfoot
+    let numcols = maximum (length aligns :
```
Checking all these things is good, defensive programming. I believe that `toLegacyTable` guarantees all lengths to be equal, so the code below could be simplified.

The use of `maximum` is ok here, as the list can never be empty. However, we generally try to avoid partial functions. How about `foldr (max . length) 0 [...]`?

tarleb commented on src/Text/Pandoc/Writers/Markua.hs:
```diff
@@ -0,0 +1,501 @@
+-- div is also used to compute the bibliography
+-- if --citeproc and --bibliography is set (div with refs id is generated),
+-- then a bibliography header level 1 with a attr list above is inserted
+blockToMarkua opts (Div attr blks) = do
```
Pattern matching might result in cleaner code. How about something like `blockToMarkua opts (Div ("refs",_,_) blks)` to handle the bibliography?

tarleb commented on src/Text/Pandoc/Writers/Markua.hs:
```diff
@@ -0,0 +1,501 @@
+    -- attr only used for bibliography
+    let attributes = attrToMarkua (citeId,[],[])
+    let biblioHeader = literal (if attrId == "refs"
+                                then "{id: bibliography}\n#Bibliography"
```
There should probably be a space after `#`, matching other headers. Could we use `blockToMarkua opts (Header ("bibliography",[],[]) "Bibliography"` to generate that part?

tarleb reviewed (commented):
Thank you, this is such a great project to do!

I've added a few first comments. Before I continue, I'd like to ask whether you'd prefer
1. a review focused on functionality, or
2. detailed review, also looking at code style and adding my (very subjective) ideas for possible improvements?

It would be helpful for us to know your time constraints, i.e., when the project has to be finished.

It's not a requirement, but do you think you may have time to maintain the writer in the future? It would be excellent to have someone to help if bugs are found by users at a later point.

jgm commented:
A basic question:  is markua so different from markdown that it makes sense to implement this in a separate module rather than as a variant of markdown?  (That is how we currently handle commonmark and plain output.)

binaarinen commented:
Thank you for your fast replies and the initial feedback.
We will start working on it right away.

***feedback***
We will have to hand in the project on the 21st of December (we know it's rather short notice). Clearly it would be great if we could get the request well enough to be merged by then and we are happy to make adjustments to improve the writer. Since we don't have unlimited time left, we would prioritise some functional feedback, but of course we would also be greatful about any other observations.


***include in markdown***
It might have been possible to include it as a variant of markdown, but because of the following differences we decided to implement it as a separate writer:
 - Markua handles linebreaks different then markdown, which affects basicly all functions.
 - Attribute lists are denoted and placed differently
 - Markua has some significant differences in it's featureset

A list of some of the main differences can also be found in the [Manual](https://leanpub.com/markua/read#markdown-diffs)

jgm commented:
> Markua handles linebreaks different then markdown, which affects basicly all functions.

We already have a switch in the markdown writer for `hard_line_breaks`.  As far as I can see, this gives the same behavior described in the Markua manual.

> Attribute lists are denoted and placed differently

This is true, but it could be handled by a new extension or switch in the markdown writer.

> Markua has some significant differences in it's featureset

Almost all of those differences are already supported by pandoc markdown extensions, e.g. list numbering features, gfm tables, definition lists, fenced code blocks.

Of those differences that aren't supported, many are irrelevant (e.g. `+` not working for bullet lists -- the markdown writer doesn't use `+` anyway for this), or the stricter standards for ATX headers, or the lack of indented code blocks (which the writer doesn't use, provided `fenced_code_blocks` is enabled).

The whitespace differences seem to be handled already by current extensions.

In Markua all URLs must be absolute; that wouldn't be enforced by the current writer, but I'm not sure how one would generate an absolute URL if the reader gives you a relative one.

One difference that does seem significance is mapping `_` to underline instead of emphasis.

The format for math is a bit different.

The blurb syntax seems new, but it's isomorphic with pandoc fenced divs and could be supported with a switch, couldn't it?  Does your writer produce this syntax, and if so from what elements in the pandoc AST?

The directives are new, but I'm not sure what in the pandoc AST should map to these (if anything).

Before merging this module, I'd need to be convinced that it makes sense to have a separate module, rather than implementing some markdown variants in the existing markdown writer (as we already do for multimarkdown, gfm, commonmark, php markdown extra, and plain).  The advantage of the latter approach is that lots of code can be reused instead of being duplicated (e.g., code for lists and pipe tables).  This can be important for maintainability in the future.

binaarinen commented:
Thank you for your feedback!
Of course we want to provide a consistent solution with the rest of the project. We are currently taking a closer look at integrating the markua part with the markdown writer and if its possible to cover all functions necessary for markua.

jgm commented:
I think that, as long as there is no Markua reader, it probably doesn't make sense to add any new markdown extensions (T.P.Extensions). The Markua variant can be handled more simply by adding a new Variant constructor in T.P.Writers.Markdown.  (See how `plain` is treated.)  Then it's just a matter of checking the variant in clauses for lists, tables, etc., and making any necessary adjustments.

binaarinen commented:
We integrated the markua writer within the markdown writer as you suggested and updated the pull request accordingly.
Please let us know if further improvements are necessary and would be happy if we could contribute to the project.

jgm commented:
Thank you -- this looks good on first glance.  I'll take a deeper look before long1
And we'll run the CI.

jgm commented:
Can you rebase this into one commit on top of current master?

jgm commented on MANUAL.txt:
```diff
@@ -252,6 +252,7 @@ header when requesting a document from a URL:
     - `markdown_mmd` ([MultiMarkdown])
     - `markdown_phpextra` ([PHP Markdown Extra])
     - `markdown_strict` (original unextended [Markdown])
+    - `markua` ([Markua])
```
This adds a Markua writer but not a reader, so we shouldn't list `markua` in the `--from` formats.

binaarinen replied:
We changed it accordingly

jgm commented on pandoc.cabal:
```diff
@@ -19,7 +19,7 @@ description:     Pandoc is a Haskell library for converting from one markup
                  - light markup formats (many variants of Markdown,
                    reStructuredText, AsciiDoc, Org-mode, Muse, Textile,
-                   txt2tags)
+                   txt2tags, Markua)
```
I'd consider Markua a "variant of Markdown," so not needing separate mention here.

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Extensions.hs:
```diff
@@ -601,6 +603,8 @@ getAllExtensions f = universalExtensions <> getAll f
     extensionsFromList
     [ Ext_smart ]
+  getAll "markua"          = extensionsFromList
+    [ Ext_citations]
```
Is this a meaningful extension for Markua?  It doesn't have automatic citation support, does it?

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -77,6 +81,25 @@ writeCommonMark opts document =
+                  enableExtension Ext_definition_lists $
+                  enableExtension Ext_smart $
+                  enableExtension Ext_footnotes $
+                    writerExtensions opts }
```
Question: does it make sense to have *any* variation in extensions for Markua, given that the whole format is specified?  E.g. would it ever make sense to add the `grid_tables` extension?  If not, then maybe best simply to hard code a list of extensions here.

binaarinen replied:
We agree, that it doesn't make much sense to change certain options. However we enabled some, that will create the correct version also used by markdown (ie. pipetables) in order not to write duplicate code.

jgm replied:
I understand; I wasn't suggesting that you not enable the extensions here.
The question is whether you should be adding them to `writerExtensions opts`, which implies that users can set extensions not on this list.  (In other words, could `writerExtensions opts` be replaced by `mempty`?  If there are no writer options it makes sense for the *user* to enable, then I think it could.)

binaarinen replied:
we changed it

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -328,7 +351,8 @@ blockToMarkdown' opts (Div attrs ils) = do
   return $
     case () of
-         _ | isEnabled Ext_fenced_divs opts &&
+         _ | variant == Markua -> contents <> blankline
```
It would be possible to use Divs with special classes to produce Markua blurbs and asides.  Would that make sense?

binaarinen replied:
We looked into the idea and it's now possible to create blurbs and asides by providing a div with class "blurb" or "aside".

jgm replied:
Have tests for this feature been added?

binaarinen replied:
we now added testcases

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -408,6 +432,9 @@ blockToMarkdown' opts b@(RawBlock f str) = do
                   "markdown_mmd", "markdown_strict"]
          -> return $ literal str <> literal "\n"
+    Markua -> do
+        report $ NoTranslation "raw content is writen out in original form and"
```
I don't understand this...  NoTranslation has a specific use, to indicate that there is no translation for a given term.
I'm not sure what you're trying to warn about here, but another LogMessage constructor should be used.

binaarinen replied:
We used NoTranslation, because it seemed to be the most fitting, since markua doesn't support raw element. It is now changed to the "renderEmpty" you suggested, which is more fitting.

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -581,27 +615,33 @@ blockToMarkdown' opts t@(Table _ blkCapt specs thead tbody tfoot) = do
   contents <- inList $ mapM (bulletListItemToMarkdown opts) items
-  return $ (if isTightList items then vcat else vsep) contents <> blankline
+  variant <- asks envVariant
+  return $ (if isTightList items || variant == Markua then vcat else vsep) 
+                  contents <> blankline
```
Can you explain this?  Is it that list items *can't* have blank lines between them in Markua?

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -581,27 +615,33 @@ blockToMarkdown' opts t@(Table _ blkCapt specs thead tbody tfoot) = do
-                       then OneParen -- commonmark only supports one paren
-                       else DefaultDelim
+  let delim' | isEnabled Ext_fancy_lists opts =
+               case variant of
+                   Markua -> if delim == TwoParens then OneParen else delim
+                   _ -> delim
+             | variant == Commonmark && --commonmark only supports one paren
+                   (delim == OneParen || delim == TwoParens) = OneParen
+             | otherwise = DefaultDelim
```
I wonder if it makes more sense to put the check for Markua outside of this guard (like the check for Commonmark below).  Then we don't need to enable `fancy_lists` extension.  This is connected to the question above, whether it makes sense to have extensions for Markua.

binaarinen replied:
The fancy_lists extension is used, because Markua supports the use of "special" enumerators (like roman numbers). Even though in this case it would be more elegant to test for Markua in a different way, this would require to add a special case in every time fancy-lists are possible. To keep the code clean and without unnecessary duplications, we opted for using the option instead as with it the Markdown part is identical to Markua.

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -581,27 +615,33 @@ blockToMarkdown' opts t@(Table _ blkCapt specs thead tbody tfoot) = do
-                               then m <> T.replicate (3 - T.length m) " "
-                               else m) markers
+  let markers' = case variant of
+                        Markua -> map (\m -> m <> T.replicate 1 " ") markers
```
`T.replicate 1 " "` == `" "`

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -581,27 +615,33 @@ blockToMarkdown' opts t@(Table _ blkCapt specs thead tbody tfoot) = do
   contents <- inList $
               zipWithM (orderedListItemToMarkdown opts) markers' items
-  return $ (if isTightList items then vcat else vsep) contents <> blankline
+  return $ (if isTightList items || variant == Markua then vcat else vsep) 
```
Same question as above on bullet lists.

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -725,19 +768,22 @@ bulletListItemToMarkdown opts bs = do
                else max (writerTabStop opts) (T.length marker + 1)
-  let start = literal marker <> sps
+  let start = case variant of
+              Markua -> literal marker
```
Don't you want at least one space after the list marker?

binaarinen replied:
Yes, the space was already passed along with the marker. We changed this slightly to improve the readability/understanding.

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -95,6 +96,13 @@ escapeText opts = T.pack . go' . T.unpack
                   , isAlphaNum x -> c : '_' : x : go xs
                 _                -> c : go cs
 
+-- Escape the escape character, as well as formatting pairs
+escapeMarkuaString :: Text -> Text
+escapeMarkuaString s = foldr (uncurry T.replace) s $ zip ["--", "**", "//",
+                                                          "^^", ",,"]
+                                                          ["~-~-", "~*~*",
+                                                           "~/~/", "~^~^",
+                                                           "~,~,", "~~"]
```
I don't think it's a good idea to use `zip` in this way -- better just to use an explicit list, where each item is paired with its escaped counterpart. (Think what would happen if one item got deleted from just one of these lists, everything would pair up with the wrong counterpart.)

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -116,9 +124,62 @@ attrsToMarkdown attribs = braces $ hsep [attribId, attribClasses, attribKeys]
+attrsToMarkua:: Attr -> Doc Text
+attrsToMarkua attributes
+    | null list = empty
+    | length list == 1 = braces $ head list
+    |  otherwise = braces $ hcat (head list : (map (\x -> ", " <> x)
```
In general, it's better to avoid using `head` (since it's a partial function and can throw runtime errors).
Why not use a case?  Match on
``` hs
  [] -> empty
  [x] -> braces x
  xs -> braces $ intercalate ", " xs
```

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -116,9 +124,62 @@ attrsToMarkdown attribs = braces $ hsep [attribId, attribClasses, attribKeys]
+                                                   (tail list)))
+        where attrId = case attributes of
+                        ("",_,_) -> []
+                        (i,_,_)  -> [literal $ T.append "id: " i]
```
More idiomatic would be `"id: " <> i`

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -116,9 +124,62 @@ attrsToMarkdown attribs = braces $ hsep [attribId, attribClasses, attribKeys]
+                                              (filter (/=(T.empty,T.empty))
+                                              (map preprocessKeyValues
+                                                   keyvalues))
+              escAttr          = mconcat . map escAttrChar . T.unpack
+              escAttrChar '"'  = literal "\""
+              escAttrChar c    = literal $ T.singleton c
```
Why not `escAttr = T.replace "\"" "\\\""`?

binaarinen replied:
We are not entirely sure where this applies, but through the process of applying your feedback we changed the entire function slightly.

jgm replied:
My point was that `escAttr = T.replace "\"" "\\\""` seems to do exactly what the three line definition above does.  Or am I wrong?

binaarinen replied:
We did not manage to simplify the function like that and still get the required results

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -116,9 +124,62 @@ attrsToMarkdown attribs = braces $ hsep [attribId, attribClasses, attribKeys]
+                               (_,_,keyvalues) -> map (\(k,v) -> escAttr k
+                                              <> ": " <>
+                                              escAttr v)
+                                              (filter (/=(T.empty,T.empty))
```
I'd probably use `(mempty, mempty)` instead of `(T.empty, T.empty)` -- for Text `mempty` is `T.empty`.

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -116,9 +124,62 @@ attrsToMarkdown attribs = braces $ hsep [attribId, attribClasses, attribKeys]
+              -- attribute gets removed
+              preprocessKeyValues :: (Text, Text) -> (Text, Text)
+              preprocessKeyValues (key,value)
+                 | value == "fig:" = (T.empty,T.empty)
+                 | T.head value == '\"' = (key, value)
```
Instead of using `(T.empty, T.empty)` for the values you're going to ignore, why not return `Maybe (Text, Text)` -- so this would be Nothing.  Then you can use `mapMaybe` on ln. 147 above.

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -116,9 +124,62 @@ attrsToMarkdown attribs = braces $ hsep [attribId, attribClasses, attribKeys]
+                 | value == "fig:" = (T.empty,T.empty)
+                 | T.head value == '\"' = (key, value)
+                 | key == "alt" ||
+                   key == "caption" = (key, T.cons '\"' (T.snoc value '\"'))
```
Even though this might be a tiny bit more efficient, for code clarity I think you can't beat `"\"" <> value <> "\""`.
Or even better define your own `inquotes`.

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -283,6 +344,7 @@ inlineToMarkdown opts (Span attrs ils) = do
              _ -> id
          $ case variant of
                 PlainText -> contents
+                Markua -> "[" <> contents <> "]" <> attrsToMarkua attrs
```
Is this right?  Markua documentation says to use backticks for spans?
(I'm no expert, just glancing at it.)

binaarinen replied:
Yes, you are correct, that must have slipped through our checks and is now corrected

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -396,60 +458,79 @@ inlineToMarkdown opts (Quoted DoubleQuote lst) = do
+  variant <- asks envVariant
+  if variant == Markua
+      then do
+        let attr = addKeyValueToAttr ("",[],[]) ("format", "latex")
+        inlineToMarkdown opts (Code attr str)
```
From the docs it looks like the preferred shorter form is `` `e=mc^2`$ ``.

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -469,6 +550,9 @@ inlineToMarkdown opts il@(RawInline f str) = do
                   "markdown_mmd", "markdown_strict"]
          -> return $ literal str
+    Markua -> do
+        report $ NoTranslation "raw content is writen out in original form and"
```
See above

jgm replied:
Why not use `renderEmpty` here? It emits the right kind of warning.

binaarinen replied:
We changed it accordingly

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -557,6 +641,8 @@ inlineToMarkdown opts lnk@(Link attr@(ident,classes,kvs) txt (src, tit)) = do
       | useAuto -> return $ literal srcSuffix
       | otherwise -> return linktext
+    Markua -> return $ "[" <> linktext <> "](" <> (literal src) <> ")" <> 
+        attrsToMarkua (addKeyValueToAttr attr ("title", tit))
```
Do you want to do this even when `tit` is empty?

binaarinen replied:
We changed it accordingly

jgm commented on test/tables.markua:
```diff
@@ -0,0 +1,58 @@
+Multiline table without column headers:
+
+|        |     |      |                                                       |
+|:------:|:----|-----:|-------------------------------------------------------|
```
Does markua support this way of indicating alignment in pipe tables with colons?  Not every implementation does.

binaarinen replied:
Yes Markua supports this [feature](https://leanpub.com/markua/read#tables)

jgm commented on test/writer.markua:
```diff
@@ -0,0 +1,718 @@
+
+[^4]: In quote.
+
+[^5]: In list.
```
Have you run this whole test output through Markua and checked to make sure everything looks right when rendered?
(Ditto with the table test?)

binaarinen replied:
Unfortunately there is currently no automatic test-tool for Markua, but we manually checked the files again, to make sure that everything is proper Markua.

binaarinen commented:
Thank you for your detailed feedback.
I'm sorry it took us a while to get to it, but we incorporated the changes to the code and rebased it into one commit.

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -327,8 +351,15 @@ blockToMarkdown' opts (Div attrs ils) = do
+     case () of
+         _ | variant == Markua -> 
+                   case () of
+                        () | classes' == ["blurb"] -> prefixed "B> " contents <> blankline
+                           | classes' == ["aside"] -> prefixed "A> " contents <> blankline
```
Blurbs can have other classes attached (e.g. in the documentation, `warning`).
So we shouldn't assume that there will be just one class, `blurb`. Instead, test to see that `"blurb"` is an element of `classes'`, and ditto for aside.

binaarinen replied:
we changed it to handle other classes aswell

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -327,8 +351,15 @@ blockToMarkdown' opts (Div attrs ils) = do
+                           | classes' == ["aside"] -> prefixed "A> " contents <> blankline
+                           -- | necessary to enable option to create a bibliography
+                           | (take 3 (T.unpack id')) == "ref" -> contents <> blankline
+                           | otherwise -> mempty
```
I don't think we should be completely omitting the contents of Divs that don't have these special classes!
Instead, just fall back to emitting the contents.

binaarinen replied:
we change it so write the contents

jgm commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -581,27 +620,33 @@ blockToMarkdown' opts t@(Table _ blkCapt specs thead tbody tfoot) = do
   contents <- inList $
               zipWithM (orderedListItemToMarkdown opts) markers' items
-  return $ (if isTightList items then vcat else vsep) contents <> blankline
+  return $ (if isTightList items then vcat else vsep) 
+                contents <> blankline
```
Let's restore the original (lack of) indentation if it fits in 80 columns.

binaarinen replied:
its back to one line

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -116,9 +122,56 @@ attrsToMarkdown attribs = braces $ hsep [attribId, attribClasses, attribKeys]
+              -- if attribute key is alt, caption, title then content
+              -- gets wrapped inside quotes
+              -- attribute gets removed
+              preprocessKeyValues :: (Text, Text) -> (Text, Text)
+              preprocessKeyValues (key,value)
+                 | key == "alt" ||
+                   key == "caption" ||
+                   key == "title" = (key, inquotes value)
```
I don't understand why these attributes are being treated differently in this way??

binaarinen replied:
The value of these keys should be in quotes, but others (ie. the format in Math) is written without, that's why they are separated

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -116,9 +122,56 @@ attrsToMarkdown attribs = braces $ hsep [attribId, attribClasses, attribKeys]
+
+-- | Add a (key, value) pair to Pandoc attr type
+addKeyValueToAttr :: Attr -> (Text,Text) -> Attr
+addKeyValueToAttr (attrlistId,classes,keyvalues) (key,value) 
+    | not (T.null key) && not (T.null value) = (attrlistId,
+                                                classes,
+                                                (key,value): keyvalues)
+    | otherwise = (attrlistId,classes,keyvalues)
```
For consistency, maybe use `ident` instead of `attrlistId`?

binaarinen replied:
we changed it to fit to the others

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
```diff
@@ -396,60 +450,77 @@ inlineToMarkdown opts (Quoted DoubleQuote lst) = do
-            (texMathToInlines DisplayMath str >>= inlineListToMarkdown opts)
+inlineToMarkdown opts (Math InlineMath str) = do
+  variant <- asks envVariant
+  if variant == Markua
+      then return $ "`" <> literal str <> "`" <> "$"
+      else
+        case writerHTMLMathMethod opts of
+          WebTeX url -> inlineToMarkdown opts
+                          (Image nullAttr [Str str] (url <> urlEncode str, str))
+          _ | isEnabled Ext_tex_math_dollars opts ->
+                return $ "$" <> literal str <> "$"
+            | isEnabled Ext_tex_math_single_backslash opts ->
+                return $ "\\(" <> literal str <> "\\)"
+            | isEnabled Ext_tex_math_double_backslash opts ->
+                return $ "\\\\(" <> literal str <> "\\\\)"
+            | otherwise -> 
+                texMathToInlines InlineMath str >>=
+                  inlineListToMarkdown opts .
+                    (if variant == PlainText then makeMathPlainer else id)
```
This is OK, but it might be nicer to avoid this nesting and just have a case statement, with the first case being:
```hs
  _ | variant == Markua -> return $ "`" etc.
```

binaarinen replied:
we changed it

binaarinen replied:
Thank You, we appreciate your fast feedback and we adjusted the code accordingly.

jgm commented on src/Text/Pandoc/Writers/Markdown/Inline.hs:
````diff
@@ -396,60 +450,77 @@ inlineToMarkdown opts (Quoted DoubleQuote lst) = do
+                  inlineListToMarkdown opts .
+                    (if variant == PlainText then makeMathPlainer else id)
+     
+inlineToMarkdown opts (Math DisplayMath str) = do
+  variant <- asks envVariant
+  if variant == Markua
+      then do
+        let attributes = attrsToMarkua (addKeyValueToAttr ("",[],[]) 
+                                                        ("format", "latex"))
+        return $ blankline <> attributes <> cr <> literal "```" <> cr 
+            <> literal str <> cr <> literal "```" <> blankline
+      else
+        case writerHTMLMathMethod opts of
+          WebTeX url -> (\x -> blankline <> x <> blankline) `fmap`
+                 inlineToMarkdown opts (Image nullAttr [Str str]
+                        (url <> urlEncode str, str))
+          _ | isEnabled Ext_tex_math_dollars opts ->
+                return $ "$$" <> literal str <> "$$"
+            | isEnabled Ext_tex_math_single_backslash opts ->
+                return $ "\\[" <> literal str <> "\\]"
+            | isEnabled Ext_tex_math_double_backslash opts ->
+                return $ "\\\\[" <> literal str <> "\\\\]"
+            | otherwise -> (\x -> cr <> x <> cr) `fmap`
+                (texMathToInlines DisplayMath str >>= inlineListToMarkdown opts)
+
````
see above on InlineMath

binaarinen replied:
we changed it

jgm reviewed (changes requested):
This is looking good.  I've made a few more comments.

Also, some small comments on your commit message.
Best if the top line says "Add a writer for Markua 1.0".
In the message, give a link to Markua's site.
Also, note the API changes using [API change]:  basically, it's just the newly exported function writeMarkua from T.P.Writers.Markdown.
You don't really need the rest, but keep the PR # and, if there's an issue in our tracker requesting Markua support, indicate that as well.

binaarinen commented:
Thank you. We appreciate your fast feedback and adjusted the code accordingly.

jgm commented on src/Text/Pandoc/Lua/Marshal/Reference.hs:
```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE LambdaCase           #-}
```
There's a bunch of stuff in this commit, like this for example, that doesn't relate to the change.
Can you rebase on current master HEAD?

binaarinen replied:
I'm sorry, we forgot to do that, it's now rebased

jgm commented:
Many thanks for your contribution to pandoc!
